staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go

```diff
@@ -156,7 +157,18 @@ func (d *memCacheClient) ServerVersion() (*version.Info, error) {
 }
 
 func (d *memCacheClient) OpenAPISchema() (*openapi_v2.Document, error) {
-	return d.delegate.OpenAPISchema()
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if d.schema == nil {
+		sch, err := d.delegate.OpenAPISchema()
```

Comment:
I didn't look at which openapi schema/spec this client is fetching. Note that CRD and aggregated apiserver can dynamically change the aggregated openapi spec we have.

I'd rather we cache the ETag returned from apiserver and utilize it in the [client](https://github.com/kubernetes/client-go/blob/e318746e79ead3763c2c1e9999ac7e7d90b725f9/discovery/discovery_client.go#L421) (I didn't look into it, but I assume this is the underlying client being used)

see related implementation for using ETag client side: https://github.com/kubernetes/kubernetes/pull/83147

Reply:
I'm not very familiar with this code, so I need some help to understand. It appears that the delegate call resolves to https://github.com/kubernetes/kubernetes/blob/c409446d41924bccb4183a87e605179db38228e5/staging/src/k8s.io/client-go/discovery/discovery_client.go#L420-L441

Are you suggesting to update this `OpenAPISchema` method to use ETags?

Reply:
Yes. Using ETags in `DiscoveryClient` makes everyone who uses `DiscoveryClient` more efficient.